src/vm_cpu.rs

```diff
@@ -96,6 +98,195 @@ pub struct VmCpuTrapState {
+            if config_flags.get_skip_match() {
+                match state {
+                    Configured(mut c) | Started(mut c) => {
+                        if config_flags.get_clear_value() {
+                            c.value = 0;
+                        }
+                        if config_flags.get_auto_start() {
+                            *state = Started(c);
+                        }
+                    }
+                    _ => {}
+                }
+            } else if matches!(state, NotConfigured) {
+                *state = if config_flags.get_auto_start() {
+                    Started(new_state)
+                } else {
+                    Configured(new_state)
+                };
+            }
+        }
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to start counters.
+    pub fn update_started_counters(
+        &mut self,
+        counter_index: u64,
+        counter_mask: u64,
+        start_flags: PmuCounterStartFlags,
+        initial_value: u64,
+    ) {
+        use PmuCounterState::*;
+        let mut i = 0u64;
+        self.counter_state
+            .iter_mut()
+            .skip(counter_index as usize)
+            .filter(|_| {
+                i += 1;
+                (counter_mask & (1 << (i - 1))) != 0
+            })
+            .all(|state| {
+                if let Configured(c) = state {
+                    if start_flags.get_init_value() {
+                        c.value = initial_value;
+                    }
+                    *state = Started(*c);
+                }
+                true
+            });
+    }
+
+    /// Updates internal state for PMU counters.
+    /// This should be called following a successful SBI call to stop counters.
+    pub fn update_stopped_counters(
+        &mut self,
+        counter_index: u64,
+        counter_mask: u64,
+        stop_flags: PmuCounterStopFlags,
+    ) {
+        use PmuCounterState::*;
+        let mut i = 0u64;
+        self.counter_state
+            .iter_mut()
+            .skip(counter_index as usize)
+            .filter(|_| {
+                i += 1;
+                (counter_mask & (1 << (i - 1))) != 0
+            })
+            .all(|state| {
+                if let Started(c) = state {
+                    if stop_flags.get_reset_flag() {
+                        *state = NotConfigured;
+                    } else {
+                        *state = PmuCounterState::Configured(*c);
+                    }
+                }
+                true
+            });
+    }
+
+    /// Returns if the PMU counter range can be started.
+    pub fn is_counter_range_startable(&self, counter_index: u64, counter_mask: u64) -> bool {
+        use PmuCounterState::*;
+        let mut i = 0u64;
+        self.counter_state
+            .iter()
+            .skip(counter_index as usize)
+            .any(|state| {
+                i += 1;
```

Comment:
I'd make an `Iterato` type that handles this "for each index starting from counter_index that's set in counter_mask" since the logic is repeated a whole bunch of times.